helper_scripts/benchmark.py

```diff
@@ -0,0 +1,135 @@
+import argparse
+import os
+from os.path import dirname
+import subprocess
+import time
+import json
+import datetime
+import shlex
+
+import requests
+from kafka import KafkaConsumer
+
+from analyze import analyze_kafka_dump
+
+
+class PopenWrapper:
+    """
+    This class is a context manager that wraps subprocess.Popen.
+    Popen waits until the created process is finished when exiting the context.
+    This wrapper additionally sends a terminate signal to the program before waiting for it to finish.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+
+    def __enter__(self):
+        self.process = subprocess.Popen(*self.args, **self.kwargs)
+        return self.process
+
+    def __exit__(self, *args):
+        self.process.terminate()
+        self.process.__exit__(*args)
+
+
+def dump_topic(topic, output_dir):
+    consumer = KafkaConsumer(topic,
+                             bootstrap_servers='kafka:9092',
+                             value_deserializer=lambda m: json.loads(m.decode('utf-8')),
+                             consumer_timeout_ms=100,
+                             auto_offset_reset='earliest')
+
+    events = [message.value for message in consumer]
+    with open(os.path.join(output_dir, topic), 'w') as file:
+        json.dump(events, file)
+
+
+def dump_kafka(output_dir):
+    kafka_dir = os.path.join(output_dir, 'kafka')
+    os.mkdir(kafka_dir)
+    topics = ['buyOffer', 'holding_cost', 'marketSituation', 'producer']
+    for topic in topics:
+        dump_topic(topic, kafka_dir)
+
+
+def save_merchant_id_mapping(output_dir):
+    merchants_info = requests.get('http://marketplace:8080/merchants').json()
+    merchant_mapping = {}
+    for merchant_info in merchants_info:
+        merchant_mapping[merchant_info['merchant_id']] = merchant_info['merchant_name']
+    with open(os.path.join(output_dir, 'merchant_id_mapping.json'), 'w') as file:
+        json.dump(merchant_mapping, file)
+
+
+def clear_containers(pricewars_dir):
+    subprocess.run(['docker-compose', 'rm', '--stop', '--force'])
+
+
+def wait_for_marketplace():
+    """
+    Send requests to the marketplace until there is a response
+    """
+    while True:
```

Comment:
That loop looks like it could run forever when we encounter some random connection problems. Are you it's doing what it should do?

Reply:
Yes, this will run forever when it is not possible to connect.
I can add a timeout and throw an exception if it cannot connect to the marketplace within this time.